### PR TITLE
Introduce retention lease actions

### PR DESCRIPTION
**Pull request by jasontedor**
This commit introduces actions for some common retention lease operations that clients need to be able to perform remotely. These actions include add/renew/remove.

Relates #37165

### CODE REVIEW

**Comment by jasontedor** [comment]
I will add the remove action to this PR after #38751 is incorporated. I will update the exceptions used in this PR after #38754 is incorporated.

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 1]
```diff
@@ -311,6 +311,11 @@ public boolean hasCompleteOperationHistory(String source, MapperService mapperSe
         return false;
     }
 
+    @Override
+    public long getMinRetainedSeqNo() {
+        throw new UnsupportedOperationException();
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: maybe return max_seq_no instead?

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseActions.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+public class RetentionLeaseActions {
+
+    static abstract class TransportRetentionLeaseAction extends TransportSingleShardAction<Request, Response> {
+
+        private Logger logger = LogManager.getLogger(getClass());
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseActions.java]: nit: final

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 3]
```diff
@@ -744,7 +744,7 @@ public Searcher acquireSearcher(String source, SearcherScope scope) throws Engin
     /**
      * Acquires a lock on the translog files and Lucene soft-deleted documents to prevent them from being trimmed
      */
-    public abstract Closeable acquireRetentionLockForPeerRecovery();
+    public abstract Closeable acquireRetentionLock();
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: ++

**Review by dnhatn — APPROVED** [review]
LGTM. I left a minor comment.

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseActions.java (hunk 2 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseActions.java]: I removed this in an iteration.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: I pushed 0518ec9a67abc8dfca25f1b70014a1c0ca79dcf1.

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**